### ISSUE TITLE
Specify how to implement organization names and logos

### ISSUE THREAD

**Issue by seanh** [hypothesis/product-backlog#515]
Spike to produce a spec for how "organizations" (previously called "publishers") should be modeled in h.

## Requirements for organizations in h

* <https://github.com/hypothesis/product-backlog/issues/497> (**Grouping groups by publisher**) requires an incremental design change to the groups menu in the client so that groups are grouped under their publishers. This is waiting on an incremental design to be posted by Dawa, but basically the menu looks like this:

  ```
  (i) SCIPUB
      Author's Notes
      Reviews
      General Discussion
      
  (i) HYPOTHESIS
      Public
      Neuro Club
      + Create a new group
  ```

Mock of target control:
![current group clustering 1](https://user-images.githubusercontent.com/5212631/37097159-a393b950-2212-11e8-9c6d-69e135dcd9d2.png)
(mock posted by Dawa)
  
  Where `(i)` stands for the icon of the organization.
  
  Related: <https://github.com/hypothesis/product-backlog/issues/502> (**Groups menu organization icons**) is a separate card about adding just the organization icons (but not names) next to the groups in the client. It may make sense to do this card first, before we have organization names, or in the end it may not make sense to do so.

![current group clustering 2](https://user-images.githubusercontent.com/5212631/37097197-b9552eea-2212-11e8-8fc4-2b37164e682a.png)
(mock posted by Dawa)
  
  So as you can see this requires each organization to have an icon and a name and some groups, and requires each group to belong to one organization (defaulting to the Hypothesis organization)

* <https://github.com/hypothesis/product-backlog/issues/492> (**The sidebar on activity pages displays the group's logo and publisher name**) is about adding these publisher names and logos to the groups activity pages. So organizations will need to be accessible to activity pages too, not just to the client.

## Out of scope for now

* **All private groups will belong to the Hypothesis organization**.

  I think we can safely say for now that all private groups (private groups currently being the only type of group that can be freely created by normal users) will belong to the Hypothesis organization. For now other organizations will contain only open and restricted groups, which must be created by Hypothesis staff admins.

* **Users won't belong to organizations**.

  I think we can safely say that for now **users** won't be directly associated with organizations. Users will just belong to groups (as they do currently) (or not even have to belong to the groups in the case of open groups) and then groups will belong to organizations. This will mean that no users have any privileges within an organization - no one can change the name or logo of an organization, or create groups within an organization. For now, all of this stuff will be done by Hypothesis staff admins. In the future we will likely want organization admins but not for now.

## Details

* Organization will have to be added as a required field to admin pages and CLIs for creating groups
* A DB migration will have to be done to move all existing groups into organizations

## Future requirements (for context only - out of scope for now)

Future requirements that I'm not sure that we have specific cards for yet but that I've seen early designs for:

* We're not implementing these yet! They're just here for context.

* Organizations may one day have their own activity pages, separate from group activity pages. For example one thing shown on an organization's activity page would be a list of all that organization's groups. Also involves changes to the navigation capabilities on activity pages to e.g. make it possible to navigate from a group's page to its organization's page.

  See also https://github.com/hypothesis/product-backlog/issues/470

* Some early activity page designs show filtering annotations by organization (an `organization:` search facet) in the search bar. Can be seen on mockups here for example: https://github.com/hypothesis/product-backlog/issues/472

* One day we will probably want to enable users to be admins of organizations. E.g. to change the organization's name or logo, or to create new groups in the organization.

## Open questions

* How does this affect the URL structure? Given that we might one day want to have organization pages, what would the URLs of these pages look like? And should the URLs of group pages change to e.g. be sub-URLs of their owning organization's pages?

**Comment by seanh** [comment]
Suggestion:

* Add an `organization` table to the DB, and corresponding `h.models.Organization` class
* `Organization` class fields will be:
  * Unique internal ID
  * Name
  * Icon (URL to image file? Or SVG stored in DB?)
  * One-many relationship to groups
* Each organization can have zero or more groups
* Each group must belong to exactly one organization
* There will be a default Hypothesis organization that the **Public** group and all existing private groups will belong to
* We will also have to create a BioPub organization and migrate BioPub's open and restricted groups into it
* All new private groups created by users go into the Hypothesis organization. No private groups in other organizations yet.
* We'll have to add organizations into the create-open-group CLI (or just delete the CLI, if we have admin pages now...)
* We'll have to add organizations into the admin pages: Create organizations (specify name, logo), edit organization name and logo, select organization when creating or editing an open or restricted group, delete organizations
* In order to do <https://github.com/hypothesis/product-backlog/issues/497> we'll have to add organization support into the API and the client. But we can do other cards (e.g. <https://github.com/hypothesis/product-backlog/issues/502> and <https://github.com/hypothesis/product-backlog/issues/492>) without any concept of organization in the API or client yet. So we only have to add organizations into the API and client when we have to.
* We'll have to add organizations in the groups menu in the top-right of activity pages as well

**Comment by seanh** [comment]
Lyza's suggested elsewhere that organizations might just be a type of group? I think this suggestion makes more sense when thinking long term: Like groups we'll want organizations to be able to have users as members, have admins, have activity pages (though not exactly the same as group activity pages. For reference [Lyza's suggestion is here](https://github.com/hypothesis/product-backlog/issues/481#issuecomment-368101851).

**Comment by seanh** [comment]
Worth noting that we can have an initial modelling of organizations that meets our immediate requirements (of showing organization names and logos on activity pages and in the groups menu) and evolve it to be able to handle later requirements later. Otherwise this can seem like a big thing to handle all at once.

**Comment by dawachan** [comment]
After speaking to the folks (authors, peer reviewers, academics) at the QDR/ATI workshop we might want to factor these user needs into group requirements:

1. users should easily be able to show/hide and control annotation privileges to groups they create. the current model assumes everything is world readable. rationale: users want to be able to expose group annotations at different times and to different audiences during the pre-post pub life cycle. 

2. users should be able to convert a group from restricted to open and open to restricted. rationale: users want to have control over who can contribute to a discussion either through annotations or replying to author’s annotations.

**Comment by seanh** [comment]
Hannah's model (from engineering meeting) (we agreed to go with this):

```
Organization
* id
* name (string)
* logo (inline svg) (nullable)

Group
* organization_id (one->many relation between organizations and groups) (required: every group must belong to exactly one organization)

Database migrations will be needed to:
* Add the organization table to the product db
* Add the organization_id column to the group table
* Create a biopub organization and relate the two existing biopub groups to it
* Create a default Hypothesis organization and relate all existing private groups and the public group to it

Admin page:
* When creating an open or restricted group you will have to select an existing organization from a drop-down menu

Private groups:
* All private groups always go into the default Hypothesis organization for now

Separate card:
* Add ability to create new organizations using the admin pages
```

(Action: @hmstepanek to break this down into cards, they can go straight into the current sprint)

**Comment by lyzadanger** [comment]
Question: What happens if an organization's last remaining group is deleted? My hunch is that it will stick around—organizations are independent of groups in that sense—but I wanted to be sure (I'm scoping how to delete groups so this popped into my head).

**Comment by hmstepanek** [comment]
Yes the organization should stick around.

**Comment by lyzadanger** [comment]
~~Also, when deleting a group, I'm hoping its `organization` relationship can be cleaned up automatically by DB constraints? If not, let me know if I need to delete that relationship manually.~~

Nevermind, the one-to-many relationship means there will just be an `organization_id` field on `group` so this is a non-thing.

**Comment by hmstepanek** [comment]
I would think it would be taken care of when it the group is deleted. It's just a reference to the organization on the group after all.

**Comment by lyzadanger** [comment]
> I would think it would be taken care of when it the group is deleted. It's just a reference to the organization on the group after all.

Yeah, I realized that within 30 seconds of posting my comment. D'oh. Moving too fast. I had already updated the comment.
